README
This is the README for the amavisd-milter(8).

NAME
     amavisd-milter -- sendmail milter for amavisd-new

SYNOPSIS
     amavisd-milter [-Bfhv] [-d debug-level] [-D delivery-care-of]
		    [-m max-conns] [-M max-wait] [-p pidfile] [-P]
		    [-q backlog] [-s socket] [-t timeout] [-S socket]
		    [-T timeout] [-w directory]

DESCRIPTION
     The amavisd-milter is a sendmail milter (mail filter) for amavisd-new
     2.4.3 and above and sendmail 8.13 and above (limited support for 8.12 is
     provided).

     Instead of older amavis-milter helper program, full amavisd-new function-
     ality is available, including adding spam and virus information header
     fields, modifying Subject, adding address extensions and removing certain
     recipients from delivery while delivering the same message to the rest.

     For more information you can visit amavisd-milter website:

	   http://amavisd-milter.sourceforge.net/

     and SourceForge project:

	   http://sourceforge.net/projects/amavisd-milter

   Options
     The options are as follows:

     -B      Uses {daemon_name} macro as a policy bank name (see POLICY BANKS
	     below).

     -d debug-level
	     Set the debug level to debug-level.  Debugging traces become more
	     verbose as the debug level increases.  Maximum is 9.

     -D delivery-care-of
	     Set AM.PDP request attribute delivery_care_of to client (default)
	     or server.  When client method is used then amavisd-milter is
	     responsible to forward the message to recipients. This method
	     doesn't allow personalized header or body modification.

	     When server method is used then amavisd-new is responsible to
	     forward the message to recipients and can provide personalized
	     header and body modification.  $forward_method in amavisd.conf
	     must point to some place willing to accept mail without further
	     checking in amavisd-new.

     -f      Run amavisd-milter in the foreground (i.e. do not daemonize).
	     Print debug messages to the terminal.

     -h      Print help page and exit.

     -m max-conns
	     Maximum concurrent amavisd connections (default 0 - unlimited
	     number of connections).  It must agree with the $max_servers
	     entry in amavisd.conf.

     -M max-wait
	     Maximum wait for connection to amavisd in seconds (default 300 =
	     5 minutes).  It must be less then sending MTA timeout for a
	     response to the final "."	that terminates a message on sending
	     MTA.  sendmail has default value 1 hour, postfix 10 minutes and
	     qmail 20 minutes.	We suggest to use less than 10 minutes.

     -p pidfile
	     Use this pid file (default /var/amavis/amavisd-milter.pid).

     -P      When amavisd-new fails mail will be passed through unchecked.

     -q backlog
	     Sets the incoming socket backlog used by listen(2).  If it is not
	     set or set to zero, the operating system default is used.

     -s socket
	     Communication socket between sendmail and amavisd-milter (default
	     /var/amavis/amavisd-milter.sock).	The protocol spoken over this
	     socket is MILTER (Mail FILTER).  It must agree with the
	     INPUT_MAIL_FILTER entry in sendmail.mc

	     The socket should be in "proto:address" format:
	     o	 {unix|local}:/path/to/file - A named pipe.
	     o	 inet:port@{hostname|ip-address} - An IPV4 socket.
	     o	 inet6:port@{hostname|ip-address} - An IPV6 socket.

     -S socket
	     Communication socket between amavisd-milter and amavisd-new
	     (default /var/amavis/amavisd.sock).  The protocol spoken over
	     this socket is AM.PDP (AMavis Policy Delegation Protocol).  It
	     must agree with the $unix_socketname entry in amavisd.conf.

	     The socket should be in "proto:address" format:
	     o	 {unix|local}:/path/to/file - A named pipe.

     -t timeout
	     sendmail connection timeout in seconds (default 600 = 10 min-
	     utes).  It must agree with the INPUT_MAIL_FILTER entry in send-
	     mail.mc and must be greater than or equal to the amavisd-new con-
	     nection timeout.  When you use other milters (especially time-
	     consuming), the timeout must be sufficient to process message in
	     all milters.

     -T timeout
	     amavisd-new connection timeout in seconds (default 600 = 10 min-
	     utes).  This timeout must be sufficient for message processing in
	     amavisd-new.  It's usually a good idea to adjust them to the same
	     value as sendmail connection timeout.

     -v      Report the version number and exit.

     -w directory
	     Set working directory (default /var/amavis).

   Limited support for sendmail 8.12
     o	 smfi_addheader() is used instead of smfi_insheader() for insheader
	 and addheader AM.PDP responses. This works well with amavisd-new
	 2.4.3 or newer.
     o	 smfi_progress() isn't called when amavisd-milter wait for amavisd-new
	 communication socket.
     o	 AM.PDP response quarantine isn't implemented.

FILES
     /var/amavis/amavisd-milter.pid
	     The default process-id file.

     /var/amavis/amavisd-milter.sock
	     The default sendmail communication socket.

     /var/amavis/amavisd.sock
	     Th default amavisd-new communication socket.

     /var/amavis
	     The default working directory.

POLICY BANKS
     If the option -B is enabled, amavisd-milter uses the value of the milter
     macro {daemon_name} as a name of the amavisd-new policy bank. Usualy this
     milter macro is set to name of the MTA.

     When remote client is authenticated, amavisd-milter uses this information
     as a name of the amavisd-new policy banks:

     SMTP_AUTH
	     Indicate that the remote client is authenticated.

     SMTP_AUTH_<MECH>
	     Remote client authentication mechanism.

     SMTP_AUTH_<MECH>_<BITS>
	     The number of bits used for the key of the symmetric cipher when
	     authentication mechanism use it.

EXAMPLES
   Configuring amavisd-new
     In amavisd.conf file change protocol and socket settings to:

	   $protocol = "AM.PDP";		      # Use AM.PDP protocol
	   $unix_socketname = "$MYHOME/amavisd.sock"; # Listen on Unix socket
	   ### $inet_socket_port = 10024;	      # Don't listen on TCP port

     Then (re)start amavisd daemon.

   Configuring sendmail
     To the sendmail.mc file add the following entries:

	   define(`confMILTER_MACROS_ENVFROM',
		   confMILTER_MACROS_ENVFROM`, r, b')
	   INPUT_MAIL_FILTER(`amavisd-milter',
		   `S=local:/var/amavis/amavisd-milter.sock,
		   F=T, T=S:10m;R:10m;E:10m')

     Then rebuild your sendmail.cf file, install it (usually to
     /etc/mail/sendmail.cf) and (re)start sendmail daemon.

   Running amavisd-milter
     This example assume that amavisd-new is running as user amavis.  It must
     agree with the entry $daemon_user in amavisd.conf.

     First create working directory:

	   mkdir /var/amavis/tmp
	   chmod 750 /var/amavis/tmp
	   chown amavis /var/amavis/tmp

     Then start amavisd-milter as non-priviledged user amavis:

	   su - amavis -c "amavisd-milter -w /var/amavis/tmp"

   Limiting maximum concurrent connections to amavisd
     To limit concurrent connections to 4 and fail after 10 minutes (10*60
     secs) of waiting run amavisd-milter with this options:

	   su - amavis -c "amavisd-milter -w /var/amavis/tmp -m 4 -M 600"

   Troubleshooting
     For troubleshooting run amavisd-milter on the foreground and set debug
     level to appropriate level:

	   su - amavis -c "amavisd-milter -w /var/amavis/tmp -f -d level"

     where debug levels are:

     1	   Not errors but unexpected states (connection abort etc).

     2	   Main states in message processing.

     3	   All amavisd-milter debug messages.

     4-9   Milter communication debugging (smfi_setdbg 1-6).

SEE ALSO
     http://amavisd-milter.sourceforge.net
     http://www.ijs.si/software/amavisd/
     http://www.milter.org/developers
     http://www.sendmail.org

AUTHORS
     This manual page was written by Petr Rehor <rx@rx.cz> and is based on
     Jerzy Sakol <jerzy.sakol@commgraf.pl> initial work.

BUGS
     A community mailing lists are available at:

	   http://sourceforge.net/mail/?group_id=138169

     Enhancements, requests and problem reports are welcome.

     If you run into problems first check the users mailing list archive
     before asking questions on the list.  It's highly likely somebody has
     already come across the same problem and it's been solved.




CHANGES
This is the CHANGELOG for amavisd-milter.

20150524:
	amavisd-milter-1.6.1:

	Bug and compatibility fixies:
	- Fixed bug when creating amavisd-new policy bank names.

20150524:
	Fixed bug when creating amavisd-new policy bank names.

	Thanks to: Christian Roessner <c@roessner-network-solutions.com>

20130519:
	amavisd-milter-1.6.0:

	New features:
	- Added new amavisd-milter option -B which passes value of
	  {daemon_name} milter macro as amavisd-new policy bank name.

	Bug and compatibility fixies:
	- Added amavisd-milter.spec for compilation with rpmbuild.
	- Fixed typo which prevents using LDFLAGS on Debian.
	- Fixed missing definition of true and false in libmilter/mfapi.h.

20130422:
	Added new amavisd-milter option -B which uses value of {daemon_name}
	milter macro as amavisd-new policy bank name.

	Thanks to: Andreas Schulze

20130125:
	Included amavisd-milter.spec for compilation with rpmbuild.

	Thanks to: Jo Rhett <jrhett@netconsonance.com>

20120727:
	Include stdbool.h before libmilter/mfapi.h in compat/compat.h because
	libmilter/mfapi.h not define true and false.

20101105:
	Fixed typo in aclocal/ax_path_milter.m4 which prevents using LDFLAGS
	on Debian.

        Thanks to: Harald Jenny <harald@a-little-linux-box.at>

20100502:
	amavisd-milter-1.5.0:

	Bug and compatibility fixies:
	- Amavisd-new 2.7.0 introduce new AM.PDP response log_id.

20100501:
	Amavisd-new 2.7.0 introduce new AM.PDP response log_id.

	Thanks to: Andreas Schulze

20100131:
	amavisd-milter-1.4.1:

	New features:
	-  Added new amavisd-milter option -P which causes that mails pass
	   through when amavisd-new fails.

	Bug and compatibility fixies:
	- Debian installs libmilter.a to /usr/lib/libmilter.
	- sendmail 8.14 introduces three new libmilter callbacks.

20091029:
	amavisd-milter(8) manual page has been improved:
	- Added link to milter documentation.
	- Added socket name format description for sendmail and amavisd-new
	  socket.

20091003:
	Added new amavisd-milter option -P which causes that mails pass through
	when amavisd-new fails.

	Thanks to: Uli Schellhaas <uli.schellhaas@memware.de>

20090803:
	Bug fix - sendmail 8.14 introduces three new callbacks.

20090601:
	Compatibility fix - Debian installs libmilter.a to /usr/lib/libmilter.

	Thanks to: Simon Hobson <linux@thehobsons.co.uk>

20081205:
	amavisd-milter-1.4.0:

	New features:
	- Added new amavisd-milter option -q which sets milter socket backlog
	  used by listen(2).
	- Local state directory (default /var/amavis) is tunable via configure
	  option:
	  --localstatedir=</DIR>	      - absolute path to local state
	  					directory.
	- Working directory (default local state directory) is tunable via
	  configure option:
	  --with-working-directory=<SUBDIR>   - subdirectory of local
						state directory.
	  --with-working-directory=</DIR>     - absolute path to working
						directory.

	Bug and compatibility fixies:
	- amavisd-milter convert CRLF to LF in mail body to support amavisd-new
	  integrated DKIM signing and verification.
	- Fixed memory leak in fts_alloc() on Solaris.
	- Solaris doesn't have dirfd().
	- Solaris 9 and older doesn't have sem_timedwait.

20081109:
	Local state directory (default /var/amavis) is tunable via configure
	option: --localstatedir=<DIR>

	Working directory (default local state directory) is tunable via
	configure option:
	--with-working-directory=<SUBDIR> subdirectory of local state directory
	--with-working-directory=</DIR>   absolute path to working directory

20081108:
	Added new amavisd-milter option -q which sets milter socket backlog
	used by listen(2).

20081029:
	Compatibility fix - Solaris 9 and older doesn't have sem_timedwait.
	Use sem_trywait if sem_timedwait isn't available.

	Thans to: Henrik Krohns <hege@hege.li>

20081029:
	Memory leak in fts_alloc() on Solaris fixed.

	Luke Mewburn wrote:
	  In examining the code, it appears that the implementation of
	  fts_alloc() can leak the struct stat fts_statp if you're not using
	  ALIGNBYTES() && ALIGN(), which I guess that Solaris doesn't have.
	  Given that the leaked buffers are 160 bytes in size (184960/1156),
	  I suspect that's sizeof(struct stat) and our culprit.

	Thanks to: Henrik Krohns <hege@hege.li>
	Submitted by: Luke Mewburn <luke@mewburn.net>

20081028:
	Compatibility fix - Solaris doesn't have dirfd().

	Thanks to: Henrik Krohns <hege@hege.li>

20080701:
	Sync fts_open.c with the NetBSD libc (upstream) fts.c, which brought
	in a few fixes, including one that looked like a potential memory leak.
	Thanks to: Luke Mewburn <luke@mewburn.net>

	Properly check -D option parameters.
	Thanks to: Henrik K <hege@hege.li>

20080310:
	Convert CRLF to LF in mail body, because sendmail send always CRLF,
	but amavisd-new requires that a message on a temporary file is in
	a native text format of a host's file system, i.e. with LF terminated
	lines on Unix or Linux.
	Thanks to: David Schweikert <david@schweikert.ch> and Mark Martinec
	<Mark.Martinec@ijs.si>

20070923:
	amavisd-milter-1.3.1

	Bug fixies:
	- Properly calculate timezone offset for synthesized Received header
	  on platforms where gmtime() and localtime() returns the pointer to
	  the same static struct.

20070923:
	Better explained -D option.

20070915:
	In mlfi.c:mlfi_envfrom() the gmtime() and localtime() call returns the
	pointer to the same static struct.  Thus pointers gt and lt are
	identical and the UTC offset is always +0000 which is bogus.
	Thanks to: Curtis Doty <curtis@greenkey.net>

20070902:
	amavisd-milter-1.3.0

	New features:
	- Implemented AM.PDP request attribute policy_bank (amavisd-new
	  2.5.0 or higher is required).  Currently SMTP_AUTH, SMTP_AUTH_<MECH>
	  and SMTP_AUTH<MECH>_<BITS> policy bank names are sent when the remote
	  client is authenticated.
	- Added new amavisd-milter option -D client|server which allow
	  personalized header or body modifications.

	Bug fixies:
	- fixed bug when amavisd communication buffer is growing.

	Compatibility:

compat/compat.h
/*
 * Copyright (c) 2005, Petr Rehor <rx@rx.cz>. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the copyright holders nor the names of its
 *    contributors may be used to endorse or promote products derived from
 *    this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNERS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
 * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
 * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
 * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * $Id: compat.h,v 1.11 2012/07/26 23:28:11 reho Exp $
 */

#ifndef _AMAVISD_COMPAT_H
#define _AMAVISD_COMPAT_H

#ifdef HAVE_CONFIG_H
# include <config.h>
#endif

#if HAVE_STDBOOL_H
# include <stdbool.h>
#else
# if ! HAVE__BOOL
#  ifdef __cplusplus
typedef bool _Bool;
#  else
typedef unsigned char _Bool;
#  endif
# endif
# define bool _Bool
# define false 0
# define true 1
# define __bool_true_false_are_defined 1
#endif

#include <errno.h>
#include <fcntl.h>
#include <libmilter/mfapi.h>
#include <libmilter/mfdef.h>
#include <limits.h>
#include <semaphore.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/param.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <sys/un.h>
#include <syslog.h>
#include <unistd.h>

#if HAVE_DIRENT_H
# include <dirent.h>
#else
# define dirent direct
# if HAVE_SYS_NDIR_H
#  include <sys/ndir.h>
# endif
# if HAVE_SYS_DIR_H
#  include <sys/dir.h>
# endif
# if HAVE_NDIR_H
#  include <ndir.h>
# endif
#endif

#if !defined(HAVE_DIRFD) && !defined(HAVE_DIRFD_AS_MACRO)
# if defined(HAVE_DIR_D_FD)
#  define dirfd(_d) ((_d)->d_fd)
# elif defined(HAVE_DIR_DD_FD)
#  define dirfd(_d) ((_d)->dd_fd)
# elif defined(HAVE_DIR___DD_FD)
#  define dirfd(_d) ((_d)->__dd_fd)
# else
#  error cannot figure out how to turn a DIR * into a fd
# endif
#endif

#if ! defined(MIN)
# define MIN(a, b)	((a) < (b) ? (a) : (b))
#endif
#if ! defined(MAX)
# define MAX(a, b)	((a) < (b) ? (b) : (a))
#endif

#if ! HAVE_DAEMON
# ifndef _PATH_DEVNULL
#  define _PATH_DEVNULL "/dev/null"
# endif
/* Run detached from the controlling terminal */
extern int     daemon(int, int);
#endif

#if HAVE_FTS_H
# include <fts.h>
#else
# include "fts_compat.h"
#endif

#if TIME_WITH_SYS_TIME
# include <sys/time.h>
# include <time.h>
#else
# if HAVE_SYS_TIME_H
#  include <sys/time.h>
# else
#  include <time.h>
# endif
#endif

#if ! HAVE_MKDTEMP
/* Make temporary directory */
extern char    *mkdtemp(char *);
#endif

#if ! HAVE_STRLCPY
/* String copy */
extern size_t	strlcpy(char *, const char *, size_t);
#endif

/* Secure socket handling */
extern ssize_t	read_sock(int, void *, size_t, long);
extern ssize_t	write_sock(int, void *, size_t, long);

#endif /* _AMAVISD_COMPAT_H */


amavisd-milter/main.c
/*
 * Copyright (c) 2005, Petr Rehor <rx@rx.cz>. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the copyright holders nor the names of its
 *    contributors may be used to endorse or promote products derived from
 *    this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNERS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
 * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
 * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
 * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * $Id: main.c,v 1.27 2013/04/22 00:36:12 reho Exp $
 */

#include "amavisd-milter.h"

#include <stdarg.h>
#include <sys/socket.h>
#include <sysexits.h>


/*
** GLOBAL VARIABLES
*/
int		daemonize = 1;
int		daemonized = 0;
int		debug_level = LOG_WARNING;
int		max_conns = 0;
int		max_wait = 5 * 60;
sem_t		max_sem_t;
sem_t	       *max_sem = NULL;
const char     *pid_file = LOCAL_STATE_DIR "/" PACKAGE ".pid";
const char     *mlfi_socket = LOCAL_STATE_DIR "/" PACKAGE ".sock";
#ifdef HAVE_SMFI_SETBACKLOG
int		mlfi_socket_backlog = 0;
#endif
long		mlfi_timeout = 600;
const char     *amavisd_socket = LOCAL_STATE_DIR "/amavisd.sock";
long		amavisd_timeout = 600;
int		ignore_amavisd_error = 0;
const char     *working_dir = WORKING_DIR;
const char     *delivery_care_of = "client";
int		policybank_from_daemon_name = 0;


/*
** USAGE - Print usage info
*/
static void
usage(const char *progname)
{
    (void) fprintf(stdout, "\nUsage: %s [OPTIONS]\n", progname);
    (void) fprintf(stdout, "Options are:\n");
    (void) fprintf(stdout, "	-B			Use daemon_name policy bank\n");
    (void) fprintf(stdout, "	-d debug-level		Set debug level\n");
    (void) fprintf(stdout, "	-D delivery		Delivery care of server or client\n");
    (void) fprintf(stdout, "	-f			Run in the foreground\n");
    (void) fprintf(stdout, "	-h			Print this page\n");
    (void) fprintf(stdout, "	-m max-conns		Maximum amavisd connections \n");
    (void) fprintf(stdout, "	-M max-wait		Maximum wait for connection in seconds\n");
    (void) fprintf(stdout, "	-p pidfile		Use this pid file\n");
    (void) fprintf(stdout, "	-P			When amavisd fails mail will be passed\n				through unchecked\n");
#ifdef HAVE_SMFI_SETBACKLOG
    (void) fprintf(stdout, "	-q backlog		Milter communication socket backlog\n");
#endif
    (void) fprintf(stdout, "	-s socket		Milter communication socket\n");
    (void) fprintf(stdout, "	-S socket		Amavisd communication socket\n");
    (void) fprintf(stdout, "	-t timeout		Milter connection timeout in seconds\n");
    (void) fprintf(stdout, "	-T timeout		Amavisd connection timeout in seconds\n");
    (void) fprintf(stdout, "	-v			Report the version and exit\n");
    (void) fprintf(stdout, "	-w directory		Set the working directory\n\n");
}


/*
** USAGEERR - Print error message, program usage and then exit
*/
static void
usageerr(const char *progname, const char *fmt, ...)
{
    char	buf[MAXLOGBUF];
    va_list	ap;

    /* Format err message */
    va_start(ap, fmt);
    (void) vsnprintf(buf, sizeof(buf), fmt, ap);
    va_end(ap);

    /* Print error message, program usage and then exit */
    (void) fprintf(stderr, "%s: %s\n", progname , buf);
    usage(progname);
    exit(EX_USAGE);
}


/*
** VERSIONINFO - Print program version info
*/
static void
versioninfo(const char *progname)
{
    (void) fprintf(stdout, "%s %s\n", progname, VERSION);
}


/*
** MAIN - Main program loop
*/
int
main(int argc, char *argv[])
{
    static	const char *args = "Bd:D:fhm:M:p:Pq:s:S:t:T:vw:";

    int		c, rstat;
    char       *p;
    const char *progname, *socket_name;
    FILE       *fp;
    struct	stat st;
    mode_t	save_umask;
    struct	sockaddr_un unix_addr;

    /* Program name */
    p = strrchr(argv[0], '/');
    if (p == NULL) {
	progname = argv[0];
    } else {
	progname = p + 1;
    }

    /* Open syslog */
    openlog(progname, LOG_PID, LOG_MAIL);

    /* Process command line options */
    while ((c = getopt(argc, argv, args)) != EOF) {
	switch (c) {
	case 'B':		/* use daemon_name policy bank */
	    policybank_from_daemon_name = 1;
	    break;
	case 'd':		/* debug level */
	    if (optarg == NULL || *optarg == '\0') {
		usageerr(progname, "option requires an argument -- %c",
		    (char)c);
	    }
	    debug_level = (int) strtol(optarg, &p, 10);
	    if (p != NULL && *p != '\0') {
		usageerr(progname, "debug level is not valid number: %s",
		    optarg);
	    }
	    if (debug_level < 0) {
		usageerr(progname, "negative debug level: %d", debug_level);
	    }
	    debug_level += LOG_WARNING;
	    break;
	case 'D':		/* delivery mechanism */
	    if (optarg == NULL || *optarg == '\0') {
		usageerr(progname, "option requires an argument -- %c",
		    (char)c);
	    }
	    if (strcmp(optarg, "client") != 0 &&
		strcmp(optarg, "server") != 0)
	    {
		usageerr(progname, "unknown delivery mechanism '%s'", optarg);
	    }
	    delivery_care_of = optarg;
	    break;
	case 'f':		/* run in foreground */
	    daemonize = 0;
	    break;
	case '?':		/* options parsing error */
	    (void) fprintf(stderr, "\n");
	case 'h':		/* help */
	    usage(progname);
	    exit(EX_OK);
	    break;
	case 'm':		/* maximum amavisd connections */
	    max_conns = (int) strtol(optarg, &p, 10);
	    if (p != NULL && *p != '\0') {
		usageerr(progname,
		    "maximum amavisd connections is not valid number: %s",
		    optarg);
	    }
	    if (max_conns < 0) {
		usageerr(progname, "negative maximum amavisd connections: %d",
		    max_conns);
	    }
	    break;
	case 'M':		/* maximum wait for connection */
	    max_wait = (int) strtol(optarg, &p, 10);
	    if (p != NULL && *p != '\0') {
		usageerr(progname,
		    "maximum wait for connection is not valid number: %s",
		    optarg);
	    }
	    if (max_wait < 0) {
		usageerr(progname, "negative maximum wait for connection: %d",
		    max_wait);
	    }
	    break;
	case 'p':		/* pid file name */
	    if (optarg == NULL || *optarg == '\0') {
		usageerr(progname, "option requires an argument -- %c",
		    (char)c);
	    }
	    pid_file = optarg;
	    break;
	case 'P':		/* when amavisd fails mail will be passed */
	    ignore_amavisd_error = 1;	/* through unchecked */
	    break;
#ifdef HAVE_SMFI_SETBACKLOG
	case 'q':		/* milter communication socket backlog */
	    mlfi_socket_backlog = (int) strtol(optarg, &p, 10);
	    if (p != NULL && *p != '\0') {
		usageerr(progname,
		    "milter communication socket backlog is not valid number: "
		    "%s", optarg);
	    }
	    if (mlfi_socket_backlog < 0) {
		usageerr(progname,
		    "negative milter communication socket backlog: %d",
		    mlfi_socket_backlog);
	    }
	    break;
#endif
	case 's':		/* milter communication socket */
	    if (optarg == NULL || *optarg == '\0') {
		usageerr(progname, "option requires an argument -- %c",
		    (char)c);
	    }
	    if (strlen(optarg) >= sizeof(unix_addr.sun_path) - 1) {
		usageerr(progname,
		    "milter communication socket name too long: %s", optarg);
	    }
	    mlfi_socket = optarg;
	    break;
	case 't':		/* milter connection timeout */
	    if (optarg == NULL || *optarg == '\0') {
		usageerr(progname, "option requires an argument -- %c",
		    (char)c);
	    }
	    mlfi_timeout = (int) strtol(optarg, &p, 10);
	    if (p != NULL && *p != '\0') {
		usageerr(progname,
		    "milter connection timeout is not valid number: %s",
		    optarg);
	    }
	    if (mlfi_timeout < 0) {
		usageerr(progname, "negative milter connection timeout: %ld",
		    mlfi_timeout);
	    }
	    break;
	case 'v':		/* version info */
	    versioninfo(progname);
	    exit(EX_OK);
	    break;
	case 'w':		/* working directory */
	    if (optarg == NULL || *optarg == '\0') {
		usageerr(progname, "option requires an argument -- %c",
		    (char)c);
	    }
	    working_dir = optarg;
	    break;
	case 'S':		/* amavisd communication socket */
	    if (optarg == NULL || *optarg == '\0') {
		usageerr(progname, "option requires an argument -- %c",
		    (char)c);
	    }
	    if (strlen(optarg) >= sizeof(unix_addr.sun_path) - 1) {
		usageerr(progname,
		    "amavisd communication socket name too long: %s", optarg);
	    }
	    amavisd_socket = optarg;
	    break;
	case 'T':		/* amavisd connection timeout */
	    if (optarg == NULL || *optarg == '\0') {
		usageerr(progname, "option requires an argument -- %c",
		    (char)c);
	    }
	    amavisd_timeout = (int) strtol(optarg, &p, 10);
	    if (p != NULL && *p != '\0') {
		usageerr(progname,
		    "amavisd connection timeout is not valid number: %s",
		    optarg);
	    }
	    if (amavisd_timeout < 0) {
		usageerr(progname, "negative amavisd connection timeout: %ld",
		    amavisd_timeout);
	    }
	    break;
	default:		/* unknown option */
	    usageerr(progname, "illegal option -- %c", (char)c);
	    break;
	}
    }

    /* Create amavisd connections semaphore */
    if (max_conns > 0) {
	if (sem_init(&max_sem_t, 0, max_conns) == -1) {
	    logmsg(LOG_ERR,
		"could not initialize amavisd connections semaphore: %s",
		strerror(errno));
	    exit(EX_SOFTWARE);
	} else {
	    max_sem = &max_sem_t;
	}
    }

    /* Check permissions on working directory */
    /* TODO: traverse working directory path */
    if (stat(working_dir, &st) != 0) {
	logmsg(LOG_ERR, "could not stat() to working directory %s: %s",
	    working_dir, strerror(errno));
	exit(EX_SOFTWARE);
    }
    if (!S_ISDIR(st.st_mode)) {
	logmsg(LOG_ERR, "%s is not directory", working_dir);
	exit(EX_SOFTWARE);
    }
    if ((st.st_mode & S_IRWXO) != 0) {
	logmsg(LOG_ERR, "working directory %s is world accessible", working_dir);
	exit(EX_SOFTWARE);
    }

    /* Configure milter */
    socket_name = NULL;
    if (mlfi_socket[0] == '/') {
	socket_name = mlfi_socket;
    }
    if (! strncmp(mlfi_socket, "unix:", 5)) {
	socket_name = mlfi_socket + 5;
    }
    if (! strncmp(mlfi_socket, "local:", 6)) {
	socket_name = mlfi_socket + 6;
    }
    if (socket_name != NULL && unlink(socket_name) != 0 && errno != ENOENT) {
	logmsg(LOG_ERR, "could not unlink old milter socket %s: %s",
	    socket_name, strerror(errno));
	exit(EX_SOFTWARE);
    }
    if (debug_level > LOG_DEBUG &&
	smfi_setdbg(debug_level - LOG_DEBUG) != MI_SUCCESS)
    {
	logmsg(LOG_ERR, "could not set milter debug level");
	exit(EX_SOFTWARE);
    }
    if (mlfi_timeout > 0 && smfi_settimeout(mlfi_timeout) != MI_SUCCESS) {
	logmsg(LOG_ERR, "could not set milter timeout");
	exit(EX_SOFTWARE);
    }
    if (smfi_setconn(mlfi_socket) != MI_SUCCESS) {
	logmsg(LOG_ERR, "could not set milter socket");
	exit(EX_SOFTWARE);
    }
    if (smfi_register(smfilter) != MI_SUCCESS) {
	logmsg(LOG_ERR, "could not register milter");
	exit(EX_SOFTWARE);
    }

    /* Unlink old pid file */
    if (pid_file != NULL) {
	if (unlink(pid_file) != 0 && errno != ENOENT) {
	    logmsg(LOG_WARNING, "could not unlink old pid file %s: %s",
		pid_file, strerror(errno));
	}
    }

    /* Run in the background */
    if (daemonize) {
	if (daemon(1, 1) != -1) {
	    daemonized = 1;
	} else {
	    logmsg(LOG_ERR, "could not fork daemon process: %s",
		strerror(errno));
	    exit(EX_OSERR);
	}
    }

    /* Connect to milter socket */
#ifdef HAVE_SMFI_SETBACKLOG
    if (mlfi_socket_backlog > 0) {
	if (smfi_setbacklog(mlfi_socket_backlog) != MI_SUCCESS) {
	    logmsg(LOG_WARNING, "could not set milter socket backlog to %d",
		mlfi_socket_backlog);
	}
    }
#endif
#ifdef HAVE_SMFI_OPENSOCKET
    if (smfi_opensocket(false) != MI_SUCCESS) {
	logmsg(LOG_ERR, "could not open milter socket %s", mlfi_socket);
        exit(EX_SOFTWARE);
    }
#endif

    /* Greetings message */
    logmsg(LOG_WARNING, "starting %s %s on socket %s", progname, VERSION,
	mlfi_socket);

    /* Create pid file */
    if (pid_file != NULL) {
	save_umask = umask(022);
	fp = fopen(pid_file, "w");
	if (fp == NULL) {
	    logmsg(LOG_WARNING, "could not create pid file %s: %s",
		pid_file, strerror(errno));
	} else {
	    (void) fprintf(fp, "%ld\n", (long) getpid());
	    if (ferror(fp)) {
		logmsg(LOG_WARNING, "could not write to pid file %s: %s",
		    pid_file, strerror(errno));
		clearerr(fp);
		(void) fclose(fp);
	    } else if (fclose(fp) != 0) {
		logmsg(LOG_WARNING, "could not close pid file %s: %s",
		    pid_file, strerror(errno));
	    }
	}
	umask(save_umask);
    }

    /* Run milter */
    if ((rstat = smfi_main()) != MI_SUCCESS) {
	logmsg(LOG_ERR, "%s failed", progname);
    } else {
	logmsg(LOG_WARNING, "stopping %s %s on socket %s", progname, VERSION,
	    mlfi_socket);
    }

    /* Unlink pid file */
    if (pid_file != NULL) {
	if (unlink(pid_file) != 0) {
	    logmsg(LOG_WARNING, "could not unlink pid file %s: %s",
		pid_file, strerror(errno));
	}
    }

    /* Destroy amavisd connections semaphore */
    if (max_sem != NULL && sem_destroy(max_sem) == -1) {
	logmsg(errno == EBUSY ? LOG_ERR : LOG_WARNING,
	    "%s: could not destroy amavisd connections semaphore: %s",
	    progname, strerror(errno));
    }

    return rstat;
}
